Initial CI setup

RyanGlScott opened the pull request:
This ports the Coq-based CI setup from `saw-script` over to `entree-specs`.

Fixes #8.

RyanGlScott committed:
Initial CI setup

This ports the Coq-based CI setup from `saw-script` over to `entree-specs`.

Fixes #8.

RyanGlScott commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,43 @@
+        shell: bash
+        run: opam repo add coq-released https://coq.inria.fr/opam/released
+
+      - name: Install Coq
+        shell: bash
+        run: opam install -y coq=${{ matrix.coq }}
+
+      - name: Install dependencies
+        shell: bash
+        run: opam install . --deps-only --yes
```
One unfortunate aspect of this setup is that we install the dependencies in two parts: the Coq-related dependencies first, then the non-Coq-related dependencies. It would be better if we could install them all at once so that we could take further advantage of parallelism. At the same time, I want to ensure that we pin the version of Coq so that it isn't upgraded without our knowledge.

Is there a way to combine these two steps so that we install all of `entree-specs`' dependencies while still pinning the Coq version? In `cabal`, this would be done via the `--constraint=coq==8.15.2` flag, but I'm unclear if `opam` has something equivalent to `--constraint`.

sauclovian-g replied:
Just change the coq constraint in the opam file. Since it doesn't actually work for 8.16 (let alone current) it's not even lying to do so :-)

RyanGlScott replied:
I agree that that would address this concern in the short term. In the long term, we could imagine testing this repo against multiple versions of Coq, similarly to how we test Haskell code against multiple releases of GHC. In that case, I still would want to provide a way of running `opam install . --deps-only` using a specific Coq version.

That being said, I don't think we need to do this now. I'll go ahead and revise the `opam` file to require `coq < 8.16` so that we don't accidentally include it in a build plan (see #1).

sauclovian-g replied:
Pragmatically that's likely to require different versions of the other bits as well, so probably the way to do it would be to symlink different opam configs, annoying as that is.

(also fwiw when I tried it in relatively current Coq it failed well before what's in #1)

sauclovian-g reviewed (approved):
Otherwise lgtm

RyanGlScott committed:
opam: Require coq < 8.16

`entree-specs` is not currently buildable with Coq 8.16 or later. See #1. As
such, we prevent these versions of Coq from being included in build plans by
excluding them at the `opam` file level.
